### PR TITLE
add indic number enable switch, enable  input wordbreak keys

### PR DESCRIPTION
**Pull request by khumnath**
learning function need to refine with this change.
*  numbers should not be wordToLearn.
* other wordBreak keys also should not be wordToLearn.

 added ```if (isWordBreakKey) {
        // Remove the word break character for learning
        wordToLearn = wordToLearn.substr(0, wordToLearn.size() - 1);
    }``` but not working i think.

### CODE REVIEW

File: src/varnam_state.cpp [hunk 1]
```diff
@@ -388,9 +392,16 @@ void VarnamState::commitText(const FcitxKeySym &key) {
         candidates->candidate(candidateSelected).text().toStringForCommit());
   }
 
-  if (isWordBreak(key)) {
-    stringToCommit = stringutils::concat(stringToCommit, getWordBreakChar(key));
-  }
+   if (isWordBreakKey) {
```
Comment (anoopmsivadas) [inline src/varnam_state.cpp]: I think its better we add another config flag to enable/disable Indic punctuation marks, and keep the default value as false.

File: src/varnam_state.cpp [hunk 2]
```diff
@@ -403,12 +414,17 @@ void VarnamState::commitText(const FcitxKeySym &key) {
...
+    
 #ifdef DEBUG_MODE
   VARNAM_INFO() << "learn word:" << stringToCommit;
 #endif
   std::string wordToLearn(stringToCommit); // [TODO] try unique_ptr<char[]>
   std::thread learnThread(varnam_learn_word, engine_->getVarnamHandle(),
                           std::move(wordToLearn), 0);
+    if (isWordBreakKey) {
+        // Remove the word break character for learning
+        wordToLearn = wordToLearn.substr(0, wordToLearn.size() - 1);
```
Comment (anoopmsivadas) [inline src/varnam_state.cpp]: Instead of this you can initialize `wordToLearn` with `stringToCommit` value before adding the punctuation char.

File: src/varnam_state.cpp [hunk 3]
```diff
@@ -370,11 +370,15 @@ void VarnamState::selectCandidate(int index) {
 void VarnamState::commitText(const FcitxKeySym &key) {
   auto candidates = ic_->inputPanel().candidateList();
   std::string stringToCommit;
+  bool isWordBreakKey = isWordBreak(key);
 
-  if (key == FcitxKey_Escape || key == FcitxKey_0 || candidates == nullptr ||
-      candidates->size() <= 1 || result_ == nullptr ||
-      varray_is_empty(result_) || lastTypedCharIsDigit) {
-
+if (key == FcitxKey_space) {
```
Comment (anoopmsivadas) [inline src/varnam_state.cpp]: Could you give your insights on why this block is added. And why do we have to handle space key specifically?.
A better approach would be to use `enableIndicNumbers`config  with `lastTypedCharIsDigit` I think.

File: .gitignore [hunk 4]
```diff
@@ -12,6 +12,7 @@ builddir/
 CMakeFiles/
 meson_options.txt
 meson.options
+CMakeLists.txt.user
```
Comment (anoopmsivadas) [inline .gitignore:15]: This can be removed right?

**Review by anoopmsivadas — CHANGES_REQUESTED** [review]

File: src/varnam_state.cpp (hunk 1 above)
Comment (khumnath) [reply src/varnam_state.cpp]: some people use English punctuation marks in typing their language even language have punctuation marks. we have to pass through punctuations with fcitx config.

File: src/varnam_state.cpp (hunk 2 above)
Comment (khumnath) [reply src/varnam_state.cpp]: i will test this.

File: .gitignore (hunk 4 above)
Comment (khumnath) [reply .gitignore:15]: i use cmake with QT Creator so I put this file to ignore list

File: src/varnam_state.cpp (hunk 3 above)
Comment (khumnath) [reply src/varnam_state.cpp]: numbers were shown in suggestion but they are not stringToCommit. i was not able to commit numbers without this change.

File: src/varnam_state.cpp (hunk 1 above)
Comment (khumnath) [reply src/varnam_state.cpp]: with new update punctuations also can be enabled or disabled

File: CMakeLists.txt [hunk 5]
```diff
@@ -25,4 +25,10 @@ add_subdirectory(icons)
 install(FILES "com.varnamproject.Fcitx5.Addon.varnamfcitx.metainfo.xml.in" 
   RENAME com.varnamproject.Fcitx5.Addon.varnamfcitx.metainfo.xml DESTINATION ${CMAKE_INSTALL_DATADIR}/metainfo)
 
-feature_summary(WHAT ALL FATAL_ON_MISSING_REQUIRED_PACKAGES)
\ No newline at end of file
+set(CMAKE_INSTALL_LIBDIR ${CMAKE_INSTALL_PREFIX}/lib)
```
Comment (khumnath) [inline CMakeLists.txt]: i created symlink to /usr/lib/x86_64-linux-gnu/fcitx5/libvarnamfcitx.so  because ldd does not found library in /usr/local/lib/fcitx5 as my system ldd configuration. if works without changes  this can be ignored

File: .gitignore (hunk 4 above)
Comment (anoopmsivadas) [inline .gitignore:15]: This can be removed right

File: CMakeLists.txt [hunk 6]
```diff
@@ -25,4 +25,10 @@ add_subdirectory(icons)
...
+install(CODE "
+  execute_process(COMMAND ${CMAKE_COMMAND} -E create_symlink 
+  ${CMAKE_INSTALL_PREFIX}/lib/fcitx5/libvarnamfcitx.so /usr/lib/x86_64-linux-gnu/fcitx5/libvarnamfcitx.so)
+")
```
Comment (anoopmsivadas) [inline CMakeLists.txt]: We're passing the INSTALL_PREFIX in CMake project setup command `-DCMAKE_INSTALL_PREFIX=/usr`. This would set the  proper install path based on the architecture. So can you remove the mentioned part and follow the build instruction as per the readme once, let me know if its not working.

File: src/varnam_config.h [hunk 7]
```diff
@@ -31,6 +31,16 @@ FCITX_CONFIGURATION(
...
     Option<bool> shouldLearnWords{this, "Learn Words", _("Learn New Words"),
                                   true};
+
+    // Enable Indic Numbers
+    Option<bool> enableIndicNumbers{this, "EnableIndicNumbers", _("Enable Indic Numbers"),
+                                    false};
+
+
+    // Enable Indic Numbers
+    Option<bool> enablePunctuation{this, "enablePunctuation", _("Enable punctuation"),
```
Comment (anoopmsivadas) [inline src/varnam_config.h]: Can you update the label to `Enable Indic Punctuation Marks`.

File: src/varnam_state.cpp [hunk 8]
```diff
@@ -368,49 +368,79 @@ void VarnamState::selectCandidate(int index) {
...
-  VARNAM_INFO() << "string to commit:" << stringToCommit << " "
-                << getWordBreakChar(key);
+    VARNAM_INFO() << "string to commit:" << stringToCommit << " "
+                  << getWordBreakChar(key);
 #endif
-  ic_->commitString(stringToCommit);
+    ic_->commitString(stringToCommit);
+if (stringToCommit.empty() || lastTypedCharIsDigit || isWordBreakKey ||
+        ic_->capabilityFlags().test(CapabilityFlag::PasswordOrSensitive) ||
+        !engine_->getConfig()->shouldLearnWords.value() || !candidateSelected) {
```
Comment (anoopmsivadas) [inline src/varnam_state.cpp]: Check if the learn word function is getting called if the last entered character is a punctuation mark.

File: src/varnam_state.cpp [hunk 9]
```diff
@@ -368,49 +368,79 @@ void VarnamState::selectCandidate(int index) {
...
+            if (getVarnamResult() && result_ && varray_length(result_) > 0) {
+                vword *result_word = static_cast<vword *>(varray_get(result_, 0));
+                if (result_word) {
+                    stringToCommit = stringutils::concat(stringToCommit, result_word->text);
+                }
+            }
+        } else {
+            // If punctuation processing is not enabled
+            // Append the punctuation character to stringToCommit
+            stringToCommit += getWordBreakChar(key);
```
Comment (anoopmsivadas) [inline src/varnam_state.cpp]: always use stringstream or similar API to concatenate std::string, otherwise it would cause unnecessary memory allocations.

**Review by anoopmsivadas — CHANGES_REQUESTED** [review]

File: src/varnam_state.cpp (hunk 9 above)
Comment (anoopmsivadas) [reply src/varnam_state.cpp]: On a second thought, don't worry about it. It wouldn't make much difference, its just two strings.

File: CMakeLists.txt (hunk 6 above)
Comment (khumnath) [reply CMakeLists.txt]: i will remove this.  i was not using install prefix when building.

File: src/varnam_state.cpp (hunk 8 above)
Comment (khumnath) [reply src/varnam_state.cpp]: i think line https://github.com/khumnath/varnam-fcitx5/blob/bce13c04f84e52332b20efd4dd2def5b362f7af0/src/varnam_state.cpp#L438 should call learning function when punctuation mark is typed.

File: .gitignore (hunk 4 above)
Comment (khumnath) [reply .gitignore:15]: yes i will remove this

File: src/varnam_config.h (hunk 7 above)
Comment (khumnath) [reply src/varnam_config.h]: ok i will do that

File: src/varnam_state.cpp (hunk 9 above)
Comment (khumnath) [reply src/varnam_state.cpp]: I did use sstream on another branch here  https://github.com/khumnath/varnam-fcitx5/blob/2cf6dc506d39a235c472b3e416bebd48d029160c/src/varnam_state.cpp#L370-L444 .
 can i pull it on dev branch?

File: src/varnam_state.cpp (hunk 8 above)
Comment (anoopmsivadas) [reply src/varnam_state.cpp]: Check whether the mentioned line is getting executed or not

File: src/varnam_state.cpp (hunk 9 above)
Comment (anoopmsivadas) [reply src/varnam_state.cpp]: We can go with std::string, no need to use std::stringstream here

File: src/varnam_state.cpp (hunk 2 above)
Comment (anoopmsivadas) [reply src/varnam_state.cpp]: Please make this change as well

**Review by anoopmsivadas — CHANGES_REQUESTED** [review]
Hi @khumnath, few comments are not addressed I believe, can you check once. Thanks!

File: .gitignore (hunk 4 above)
Comment (khumnath) [reply .gitignore:15]: this is removed.

File: CMakeLists.txt (hunk 6 above)
Comment (khumnath) [reply CMakeLists.txt]: this was my personnel thing and removed now.

File: src/varnam_state.cpp (hunk 8 above)
Comment (khumnath) [reply src/varnam_state.cpp]: When I type a word like ‘प्रवन्ध’ following with word break key or a digit, the word is learned. I haven’t tested it in other situations, but it should work as expected based on this code.

File: src/varnam_state.cpp (hunk 9 above)
Comment (khumnath) [reply src/varnam_state.cpp]: if we can use std::string then this is solved i think.

**Comment by khumnath** [comment]
> Hi @khumnath, few comments are not addressed I believe, can you check once. Thanks!

I have commented on previous requests as I made changes, but I’m not sure where you are expecting my comment to address your request.

File: src/varnam_state.cpp [hunk 10]
```diff
@@ -368,49 +368,79 @@ void VarnamState::selectCandidate(int index) {
...
+      wordToLearn = stringToCommit;
 #ifdef DEBUG_MODE
-  VARNAM_INFO() << "string to commit:" << stringToCommit << " "
-                << getWordBreakChar(key);
+    VARNAM_INFO() << "string to commit:" << stringToCommit << " "
+                  << getWordBreakChar(key);
 #endif
-  ic_->commitString(stringToCommit);
+    ic_->commitString(stringToCommit);
+if (stringToCommit.empty() || lastTypedCharIsDigit || isWordBreakKey ||
```
Comment (anoopmsivadas) [inline src/varnam_state.cpp]: Here we're returning from the function if its word break key right! so the learning part will be skipped. I haven't tested this though, not sure if I missed anything else either. Please check this once.

Comment (khumnath) [reply src/varnam_state.cpp]: may be i did not tested that. can you check this whole ```commitText``` ? 
````
void VarnamState::commitText(const FcitxKeySym &key) {
    auto candidates = ic_->inputPanel().candidateList();
    std::string stringToCommit;
    std::string wordToLearn;
    bool isWordBreakKey = isWordBreak(key);
    bool enablePunctuation = engine_->getConfig()->enablePunctuation.value();

    if (isWordBreakKey) {
        if (candidates && candidates->cursorIndex() >= 0) {
            stringToCommit = candidates->candidate(candidates->cursorIndex()).text().toStringForCommit();
        } else {
            stringToCommit = preedit_.toStringForCommit();
        }

        if (enablePunctuation) { // If punctuation processing is enabled
            buffer_.clear(); // Clear buffer
            buffer_.push_back(getWordBreakChar(key)[0]); // Add word break char to buffer
            
            // Concatenate result word if available
            if (getVarnamResult() && result_ && varray_length(result_) > 0) {
                vword *result_word = static_cast<vword *>(varray_get(result_, 0));
                if (result_word) {
                    stringToCommit = stringutils::concat(stringToCommit, result_word->text);
                }
            }
        } else {
            // If punctuation processing is not enabled
            // Append the punctuation character to stringToCommit
            stringToCommit += getWordBreakChar(key);
            // Set wordToLearn to the base candidate text without punctuation
            wordToLearn = candidates->candidate(candidateSelected).text().toStringForCommit();
        }
    } else if (key == FcitxKey_Escape || key == FcitxKey_0 ||
               !candidates || candidates->size() <= 1 || !result_ || varray_is_empty(result_)) {
        // Handle escape or special keys
        stringToCommit.assign(preedit_.toStringForCommit());
        candidateSelected = 0;
    } else if ((candidates->cursorIndex() <= 0) && !candidateSelected) {
        // Handle the case where no candidate is selected
        vword *first_result = static_cast<vword *>(varray_get(result_, 0));
        if (first_result) {
            stringToCommit.assign(first_result->text);
            candidateSelected = 1;
        }
    } else {
        // Handle regular candidate selection
        stringToCommit.assign(candidates->candidate(candidateSelected).text().toStringForCommit());
    }

    wordToLearn = stringToCommit;

    ic_->commitString(stringToCommit);

    // Early exit if learning should be skipped
    if (stringToCommit.empty() || 
        ic_->capabilityFlags().test(CapabilityFlag::PasswordOrSensitive) ||
        !engine_->getConfig()->shouldLearnWords.value() || !candidateSelected) {
        reset();
        return;  // Exit early if learning should be skipped
    }

#ifdef DEBUG_MODE
    VARNAM_INFO() << "Original string to commit:" << stringToCommit;  // Log original string
#endif

    // Adjust wordToLearn by removing only the word break character if `isWordBreakKey` is true
    if (isWordBreakKey && enablePunctuation) {
        wordToLearn = wordToLearn.substr(0, wordToLearn.size() - 1);  // Remove the word break character
    }

    // Remove the last character if it's a digit
    if (lastTypedCharIsDigit) {
        wordToLearn = wordToLearn.substr(0, wordToLearn.size() - 1);  // Remove last digit
    }

#ifdef DEBUG_MODE
    VARNAM_INFO() << "Learn word after adjustments:" << wordToLearn;  // Log the adjusted word to learn
#endif

    // Start the learning thread with the word to learn (without word break or digit)
    std::thread learnThread(varnam_learn_word, engine_->getVarnamHandle(),
                            std::move(wordToLearn), 0);
    learnThread.detach();

    reset();  // Reset after learning the word
}
```

Comment (khumnath) [reply src/varnam_state.cpp]: check new push if that solves this issue

**Review by anoopmsivadas — COMMENTED** [review]
LGTM :) Made few changes to fix the learning part and did a quick clean up.

@khumnath, Thanks for bearing with me 😅
